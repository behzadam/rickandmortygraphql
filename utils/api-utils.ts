import axios from 'axios'
import { print, DocumentNode } from 'graphql'

// To fix nested data object: response.data.data
axios.interceptors.response.use(
  function (response) {
    return response.data
  },
  function (error) {
    return Promise.reject(error)
  }
)

// You can put it in .env
const endpoint = 'https://rickandmortyapi.com/graphql'

/**
 * We have to use POST method here.
 * @param query - generated DocumentNode by 'yarn graphql:codegen'
 * @param variables - query params
 * @returns Promise<T>
 */
export const fetcher = <T>(
  query: DocumentNode,
  variables: Record<string, string> = {}
): Promise<T> => {
  return new Promise<T>((resolve, reject) => {
    axios
      .post<T>(endpoint, { query: print(query), variables })
      .then((response) => resolve(response.data))
      .catch((error) => reject(error))
  })
}
