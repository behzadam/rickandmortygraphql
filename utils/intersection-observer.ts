import { RefObject, useEffect } from 'react'

type Props = {
  root?: RefObject<HTMLElement>
  target?: RefObject<HTMLElement>
  rootMargin?: string
  threshold?: number
  enabled: boolean
  onIntersect: () => void
}

export default function useIntersectionObserver({
  root,
  target,
  rootMargin = '0px',
  threshold = 1.0,
  enabled = true,
  onIntersect,
}: Props) {
  useEffect(() => {
    if (!enabled) {
      return
    }

    const observer = new IntersectionObserver(
      (entries) =>
        entries.forEach((entry) => entry.isIntersecting && onIntersect()),
      {
        root: root && root.current,
        rootMargin,
        threshold,
      }
    )

    const el = target && target.current
    if (!el) {
      return
    }

    observer.observe(el)

    return () => {
      observer.unobserve(el)
    }
  }, [target, enabled])
}
