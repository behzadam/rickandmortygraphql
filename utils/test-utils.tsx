import React, { ReactElement, useState } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { QueryClient, QueryClientProvider } from 'react-query'

type Props = {
  children?: React.ReactElement
}

const Providers = ({ children }: Props): JSX.Element => {
  const [queryClient] = useState(
    () =>
      new QueryClient({
        defaultOptions: {
          queries: {
            refetchOnWindowFocus: false,
          },
        },
      })
  )
  return (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: Providers, ...options })

// re-export
export * from '@testing-library/react'
export { customRender as render }
