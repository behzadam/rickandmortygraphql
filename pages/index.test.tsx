import React from 'react'

import { render } from '../utils/test-utils'
import Home from './index'

describe('HomePage', () => {
  it('should render the heading', () => {
    const { asFragment } = render(<Home />)
    const firstRender = asFragment()
    expect(firstRender).toMatchSnapshot(asFragment())
  })
})
