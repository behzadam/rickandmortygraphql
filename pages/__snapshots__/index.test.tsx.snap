// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage should render the heading 1`] = `Object {}`;
